Add Network ACL support in EC2 controller

nishant221 opened the pull request:
Issue (https://github.com/aws-controllers-k8s/community/issues/1559)

Description of changes:
Added support in EC2 controller for adding/updating/creating NACL.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

RedbackThomson reviewed (changes requested):
Great job with this. Mostly some nitpicks. Would you mind also adding a test case, which uses the custom hook code?

nnbu commented on apis/v1alpha1/generator.yaml:
```diff
@@ -17,6 +17,9 @@ ignore:
     - CreateNatGatewayInput.ClientToken
     - CreateNatGatewayInput.DryRun
     - CreateNatGatewayInput.TagSpecifications
+    - CreateNetworkAclInput.ClientToken
```
`CreateNetworkAcl` does not seem to have `ClientToken`. Do we need this line here? https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_CreateNetworkAcl.html

nishant221 replied:
Removed.

nnbu commented on apis/v1alpha1/generator.yaml:
```diff
@@ -514,6 +519,46 @@ resources:
+    fields:
+      # support EC2-VPC only
+      Id:
+        #is_primary_key: true
```
nit: please remove the commented line `#is_primary_key: true`

nishant221 replied:
Removed.

nnbu commented on config/crd/bases/ec2.services.k8s.aws_networkaclentries.yaml:
```diff
@@ -0,0 +1,182 @@
+  annotations:
+    controller-gen.kubebuilder.io/version: v0.9.2
+  creationTimestamp: null
+  name: networkaclentries.ec2.services.k8s.aws
```
This file does not seem to be generated by code-generator, unless you add `netoworkAclEntry` resource in generator.yaml. Is this residual file to be removed?

nishant221 replied:
Yes. This file is a residual one. Thanks for notifying. Removed the file.

nnbu commented on apis/v1alpha1/ack-generate-metadata.yaml:
```diff
@@ -1,13 +1,13 @@
+  build_date: "2023-06-19T11:13:03Z"
+  build_hash: e9b68590da73ce9143ba1e4361cebdc1d876c81e
+  go_version: go1.19
+  version: v0.26.1-7-ge9b6859-dirty
```
Could you update the tags in `code-generator` repo and rebuild this? `version` should not be `dirty`.

nishant221 replied:
Fixed this.

nishant221 commented on apis/v1alpha1/generator.yaml:
```diff
@@ -514,6 +519,46 @@ resources:
+      custom_method_name: customUpdateNetworkAcl
+
+
+
```
Removed.

nishant221 commented on pkg/resource/network_acl/hooks.go:
```diff
@@ -0,0 +1,479 @@
+import (
+	//svcapitypes "github.com/aws-controllers-k8s/ec2-controller/apis/v1alpha1"
+	//ackcompare "github.com/aws-controllers-k8s/runtime/pkg/compare"
+	//ackrtlog "github.com/aws-controllers-k8s/runtime/pkg/runtime/log"
```
Removed.

nishant221 commented on pkg/resource/network_acl/hooks.go:
```diff
@@ -0,0 +1,479 @@
+	ctx context.Context,
+	desired *resource,
+	latest *resource,
+
```
Removed.

nishant221 commented on pkg/resource/network_acl/hooks.go:
```diff
@@ -0,0 +1,479 @@
+		}
+	}
+
+	// // Checking latest for entries which has the same rule number as the entry in toAdd
```
Changed.

nishant221 commented on pkg/resource/network_acl/hooks.go:
```diff
@@ -0,0 +1,479 @@
+	}
+
+	for _, entry := range toAdd {
+		rlog.Debug("adding entries to nacl")
```
Removed `rlog.Debug` lines.

nishant221 commented on pkg/resource/network_acl/hooks.go:
```diff
@@ -0,0 +1,479 @@
+) error {
+	if err := rm.syncRules(ctx, r, nil); err != nil {
+		return err
+	}
```
Encapsulated.

nishant221 commented on pkg/resource/network_acl/hooks.go:
```diff
@@ -0,0 +1,479 @@
+
+	for _, desiredEntry := range desired.ko.Spec.Entries {
+
+		if *((*desiredEntry).RuleNumber) == int64(32767) {
```
Moved the value to a variable.

nishant221 commented:
Incorporated some of the feedbacks. Working on rest of them and on adding the test cases.

nishant221 commented:
@RedbackThomson Added e2e-tests as well. Will it be possible to add `ok-to-test` on this PR ?

nishant221 commented on pkg/resource/network_acl/hooks.go:
```diff
@@ -0,0 +1,479 @@
+}
+
+// computeTagsDelta returns tags to be added and removed from the resource
+func computeTagsDelta(
```
Modified as per the suggestion.

nishant221 commented on apis/v1alpha1/network_acl.go:
```diff
@@ -0,0 +1,82 @@
+	IsDefault *bool `json:"isDefault,omitempty"`
+	// The ID of the network ACL.
+	// +kubebuilder:validation:Optional
+	NetworkACLID *string `json:"networkACLID,omitempty"`
```
Have a small doubt on this one. Is this what we are referring to https://github.com/aws-controllers-k8s/ec2-controller/pull/140/files#diff-e8312348bafc4b9c54ba32ead53d9e158a234f574eeebe73c846c20d3548e6b0R524-R527 ?

nnbu replied:
I think, @RedbackThomson is referring to [renaming](https://aws-controllers-k8s.github.io/community/docs/contributor-docs/code-generator-config/#renames-renaming-things) the field using `rename` directive. 

The code you pointed is used for [printing](https://aws-controllers-k8s.github.io/community/docs/contributor-docs/code-generator-config/#print-controlling-a-fields-output-as-printer-columns-in-kubectl-get).

RedbackThomson replied:
@nnbu Is correct. I don't think it's necessary to print the ID (although we could, too), but I'd at least like to see it renamed.

nishant221 replied:
@RedbackThomson  As suggested, renamed `NetworkAclID` to `ID`.

nishant221 commented:
@RedbackThomson Incorporated all the feedbacks and added e2e tests as well. 
There is only one feedback that I have added a follow-up comment on to understand it better.

Kindly have a look.
Thanks

RedbackThomson commented:
/ok-to-test

RedbackThomson commented on test/e2e/tests/test_network_acl.py:
```diff
@@ -0,0 +1,282 @@
+        ec2_validator.assert_network_acl(network_acl_id, exists=False)
+
+
+    @pytest.mark.resource_data({'tag_key': 'initialtagkey', 'tag_value': 'initialtagvalue'})
```
This is interesting. Why did you choose this method over simply hardcoding it?

nishant221 replied:
Yes, we can hardcode it as well. Tried to follow code uniformity with other resources test cases (e.g. https://github.com/aws-controllers-k8s/ec2-controller/blob/main/test/e2e/tests/test_nat_gateway.py#L155 )

RedbackThomson replied:
I don't think it adds anything to the tests, for now, and adds complexity into the fixture logic. Could we please hard code it instead?

nishant221 replied:
HardCoded.

nnbu commented on apis/v1alpha1/generator.yaml:
```diff
@@ -514,6 +518,41 @@ resources:
+        print:
+          path: Status.networkACLID
+          name: ID
+      Associations:
+        custom_field:
+          list_of: NetworkAclAssociation
```
I don't any logic which kicks in if `Association` is specified in the spec. Do you plan to add this support in a future PR?

nishant221 replied:
@nnbu Yes, planning on covering it up in a follow-up PR.

nishant221 replied:
@nnbu Added support for `Association` as well with the commit : https://github.com/aws-controllers-k8s/ec2-controller/pull/140/commits/58a814bf7c14540958728fd08a4e6667ccdaf692

nishant221 commented:
Working on fixing e2e test errors and incorporating leftover feedbacks.

nishant221 commented:
e2e-tests are now passing.

nishant221 commented:
@RedbackThomson Incorporated all the feedbacks and e2e tests are also now passing. Kindly have a look at your convenience .
Thanks

nishant221 commented:
@RedbackThomson Added support for `Association` as well with the commit : https://github.com/aws-controllers-k8s/ec2-controller/pull/140/commits/58a814bf7c14540958728fd08a4e6667ccdaf692
Kindly have a look at your convenience. 



We use NACL in our infra and it will be of great help for us if this can be reviewed/merged :)

RedbackThomson commented on pkg/resource/network_acl/hooks.go:
```diff
@@ -0,0 +1,658 @@
+	return nil
+}
+
+// // sdkTags converts *svcapitypes.Tag array to a *svcsdk.Tag array
+func (rm *resourceManager) sdkTags(
+
+	tags []*svcapitypes.Tag,
+
+) (sdktags []*svcsdk.Tag) {
```
nit: You have 2 comment blocks, and please remove the spacing around the parameters

nishant221 replied:
Fixed.

RedbackThomson commented on pkg/resource/network_acl/hooks.go:
```diff
@@ -0,0 +1,658 @@
+
+// compareTags is a custom comparison function for comparing lists of Tag
+// structs where the order of the structs in the list is not important.
+func compareTags(
```
We aren't using this anymore? Please delete

nishant221 replied:
We are using it `delta.go`.

RedbackThomson commented on pkg/resource/network_acl/hooks.go:
```diff
@@ -0,0 +1,658 @@
+	updated = rm.concreteResource(desired.DeepCopy())
+
+	if delta.DifferentAt("Spec.Entries") {
+		if err := rm.syncRules(ctx, desired, latest); err != nil {
```
Any particular reason you didn't call this `syncEntries`?

RedbackThomson commented on pkg/resource/network_acl/hooks.go:
```diff
@@ -0,0 +1,658 @@
+	return r.ko.Status.ID == nil
+}
+
+func (rm *resourceManager) createRules(
```
Same as above - why not `createEntries`?

nishant221 replied:
Yes, `createEntries ` makes more sense. Changed.

RedbackThomson commented on pkg/resource/network_acl/hooks.go:
```diff
@@ -0,0 +1,658 @@
+	ctx context.Context,
+	r *resource,
+) error {
+	if r.ko.Spec.Entries != nil {
```
This should be checking for `r.ko.Spec.Associations`

nishant221 replied:
Changed.

RedbackThomson commented on pkg/resource/network_acl/hooks.go:
```diff
@@ -0,0 +1,658 @@
+	if latest != nil {
+		for _, association := range latest.ko.Spec.Associations {
+			if association.NetworkACLAssociationID != nil {
+				latest_associations[*association.NetworkACLAssociationID] = "NetworkACLAssociationID"
+			}
+			if association.SubnetID != nil {
+				latest_associations[*association.SubnetID] = "SubnetID"
```
Please move magic strings into consts at the top of the file

nishant221 replied:
Moved.

RedbackThomson commented on pkg/resource/network_acl/hooks.go:
```diff
@@ -0,0 +1,658 @@
+	}
+	to_Add := lo.OmitByKeys(desired_associations, lo.Keys(latest_associations))
+	included_subnets := lo.PickByKeys(associationid_subnet, lo.Keys(desired_associations))
+	y := lo.OmitByKeys(latest_associations, lo.Keys(desired_associations))
```
Please rename this variable to something more understandable

nishant221 replied:
Changed.

RedbackThomson commented on pkg/resource/network_acl/hooks.go:
```diff
@@ -0,0 +1,658 @@
+			}
+		}
+	}
+	to_Add := lo.OmitByKeys(desired_associations, lo.Keys(latest_associations))
```
nit: Could you add a little comment above this just explaining what all of these omits and picks achieve?

nishant221 replied:
Added.

RedbackThomson commented on pkg/resource/network_acl/hooks.go:
```diff
@@ -0,0 +1,658 @@
+			}
+			input.NetworkAclId = desired.ko.Status.ID
+			if len(dna_output.NetworkAcls) != 1 {
+				return errors.New("unexpected output from describenetworkacls for the given subnet")
```
```suggestion
return errors.New("unexpected output from DescribeNetworkAcls for the given subnet")
```

nishant221 replied:
Changed.

RedbackThomson replied:
Also don't see this

RedbackThomson commented on pkg/resource/network_acl/hooks.go:
```diff
@@ -0,0 +1,658 @@
+			input.NetworkAclId = desired.ko.Status.ID
+			if len(dna_output.NetworkAcls) != 1 {
+				return errors.New("unexpected output from describenetworkacls for the given subnet")
+			} else {
```
nit: Don't need `else` after a `return`

nishant221 replied:
Removed.

RedbackThomson commented on pkg/resource/network_acl/hooks.go:
```diff
@@ -0,0 +1,658 @@
+				},
+			}
+			dna_output, err := rm.sdkapi.DescribeNetworkAclsWithContext(ctx, dna_input)
+			rm.metrics.RecordAPICall("DESCRIBE", "DescribeNetworkAcls", err)
```
```suggestion
rm.metrics.RecordAPICall("READ_MANY", "DescribeNetworkAcls", err)
```

nishant221 replied:
Changed.

RedbackThomson replied:
Don't see this change

nishant221 replied:
Apologies for missing this one. Pushed the required changes now.

RedbackThomson commented on pkg/resource/network_acl/hooks.go:
```diff
@@ -0,0 +1,658 @@
+		},
+	}
+	default_nacl, err := rm.sdkapi.DescribeNetworkAclsWithContext(ctx, x)
+	rm.metrics.RecordAPICall("DESCRIBE", "DescribeNetworkAcls", err)
```
```suggestion
rm.metrics.RecordAPICall("READ_MANY", "DescribeNetworkAcls", err)
```

nishant221 replied:
Changed.

RedbackThomson commented on pkg/resource/network_acl/hooks.go:
```diff
@@ -0,0 +1,658 @@
+		return err
+	}
+	if default_nacl == nil {
+		return errors.New("could not determine")
```
Please finish this error message

nishant221 replied:
Modified.

RedbackThomson commented on pkg/resource/network_acl/hooks.go:
```diff
@@ -0,0 +1,658 @@
+		return errors.New("could not determine")
+	}
+	for rid, rtype := range to_Delete {
+		input := &svcsdk.ReplaceNetworkAclAssociationInput{}
```
I feel like this logic is very long for 1 method - perhaps we could break it up into `upsertNewAssociations` and `deleteOldAssociations`?

nishant221 replied:
Function splitted into `upsertNewAssociations` and `deleteOldAssociations`

RedbackThomson commented on pkg/resource/network_acl/hooks.go:
```diff
@@ -0,0 +1,658 @@
+
+}
+
+func toStrPtr(str string) *string {
+	return &str
+}
```
Please use `lo.ToPtr`

nishant221 replied:
Changed `toStrPtr` to `lo.ToPtr`

RedbackThomson commented on pkg/resource/network_acl/hooks.go:
```diff
@@ -0,0 +1,658 @@
+	"github.com/samber/lo"
+)
+
+var defaultRuleNumber = 32767
```
nit: Use proper case for consts:
```suggestion
var DefaultRuleNumber = 32767
```

nishant221 replied:
Changed.

RedbackThomson commented on pkg/resource/network_acl/hooks.go:
```diff
@@ -0,0 +1,658 @@
+	toDelete := []*svcapitypes.NetworkACLEntry{}
+	toUpdate := []*svcapitypes.NetworkACLEntry{}
+
+	ruleNumber := make(map[int]bool)
+
+	for _, entry := range desired.ko.Spec.Entries {
+		if value, present := ruleNumber[int(*entry.RuleNumber)]; present {
+			if value == (*entry.Egress) {
+				return errors.New("multple rules with the same rule number and Egress in the desired spec")
+			}
+		} else {
+			ruleNumber[int(*entry.RuleNumber)] = (*entry.Egress)
+		}
+	}
```
It seems this logic is just checking for uniqueness? We can use `lo.UniqBy` instead

nishant221 replied:
Modified the logic to use `lo.UniqBy`

RedbackThomson commented on pkg/resource/network_acl/hooks.go:
```diff
@@ -0,0 +1,658 @@
+
+	res.NetworkAclId = r.ko.Status.ID
+	_, err = rm.sdkapi.CreateNetworkAclEntryWithContext(ctx, res)
+	rm.metrics.RecordAPICall("CREATE", "CreateNetworkAclEntry", err)
```
```suggestion
rm.metrics.RecordAPICall("UPDATE", "CreateNetworkAclEntry", err)
```

nishant221 replied:
Changed.

RedbackThomson commented on pkg/resource/network_acl/hooks.go:
```diff
@@ -0,0 +1,658 @@
+
+	res.NetworkAclId = r.ko.Status.ID
+	_, err = rm.sdkapi.ReplaceNetworkAclEntryWithContext(ctx, res)
+	rm.metrics.RecordAPICall("REPLACE", "ReplaceNetworkAclEntry", err)
```
```suggestion
rm.metrics.RecordAPICall("UPDATE", "ReplaceNetworkAclEntry", err)
```

nishant221 replied:
Changed.

RedbackThomson commented on pkg/resource/network_acl/hooks.go:
```diff
@@ -0,0 +1,658 @@
+
+	res.NetworkAclId = r.ko.Status.ID
+	_, err = rm.sdkapi.DeleteNetworkAclEntryWithContext(ctx, res)
+	rm.metrics.RecordAPICall("DELETE", "DeleteNetworkAclEntry", err)
```
```suggestion
rm.metrics.RecordAPICall("UPDATE", "DeleteNetworkAclEntry", err)
```

nishant221 replied:
Changed.

RedbackThomson commented on test/e2e/tests/helper.py:
```diff
@@ -85,6 +85,20 @@ def assert_route(self, route_table_id: str, gateway_id: str, origin: str, exists
+        try:
+            aws_res = self.ec2_client.describe_network_acls(NetworkAclIds=[network_acl_id])
+            entries = aws_res["NetworkAcls"][0]["Entries"]
+            print("mydata:")
+            print(entries)
```
Please remove

nishant221 replied:
Cleaned.

RedbackThomson commented on test/e2e/tests/test_network_acl.py:
```diff
@@ -0,0 +1,278 @@
+def simple_network_acl(request):
+    resource_name = random_suffix_name("network-acl-test", 24)
+    resources = get_bootstrap_resources()
+    #igw_id = resources.SharedTestVPC.public_subnets.route_table.internet_gateway.internet_gateway_id
+
+    replacements = REPLACEMENT_VALUES.copy()
+    replacements["NETWORK_ACL_NAME"] = resource_name
+    replacements["VPC_ID"] = resources.SharedTestVPC.vpc_id
+    replacements["CIDR_BLOCK"] = "192.168.1.0/24"
+
+
+
+    #replacements["TAG_KEY"] = "Name"
+    #replacements["TAG_VALUE"] = resource_name
```
Please clean

nishant221 replied:
Cleaned.

RedbackThomson commented on test/e2e/tests/test_network_acl.py:
```diff
@@ -0,0 +1,278 @@
+        (ref, cr) = simple_network_acl
+
+        resource = k8s.get_resource(ref)
+        print("mydata")
+        print(resource)
```
Please clean

nishant221 replied:
Cleaned.

RedbackThomson commented on test/e2e/tests/test_network_acl.py:
```diff
@@ -0,0 +1,278 @@
+
+        # Check system and user tags exist for networkAcl resource
+        network_acl = ec2_validator.get_network_acl(resource_id)
+        print(network_acl)
```
Clean

nishant221 replied:
Done.

RedbackThomson commented on test/e2e/tests/test_network_acl.py:
```diff
@@ -0,0 +1,278 @@
+
+        # Patch the networkAcl resource, deleting the tags
+        updates = {
+                "spec": {"tags": []},
```
Maybe we can add new associations and entries to test their update code path?

nishant221 replied:
Added required associations test as well.

nishant221 commented:
@RedbackThomson Thank you for reviewing this. I have incorporated all the mentioned feedbacks and suggestions. Kindly have a look at your convenience. 
Thanks

A-Hilaly commented on pkg/resource/network_acl/sdk.go:
```diff
@@ -0,0 +1,651 @@
+		}
+		found = true
+		break
+	}
```
Not sure but maybe we also also need to set `status.ID` in the above? in case we try to adopt this resource, `requiredFieldsMissingFromReadManyInput` will always return false causing the `skdFidn` to return `ack.NotFound`.

nishant221 replied:
Not sure on this. I tried to adopt a nacl resource via below and it works :
```
apiVersion: services.k8s.aws/v1alpha1
kind: AdoptedResource
metadata:
  name: test-nacl-adopted
spec:
  aws:
    nameOrID: <nacl_id>
  kubernetes:
    group: ec2.services.k8s.aws
    kind: NetworkACL
```

```
Status:
  Ack Resource Metadata:
    Owner Account ID:  <account-id>
    Region:            us-west-2
  Conditions:
    Last Transition Time:  2023-07-07T10:02:27Z
    Message:               Resource synced successfully
    Reason:
    Status:                True
    Type:                  ACK.ResourceSynced
  Id:                      <nacl_id>
  Is Default:              false
  Owner ID:                <account-id>
Events:                    <none>
```

A-Hilaly commented on pkg/resource/network_acl/hooks.go:
```diff
@@ -0,0 +1,677 @@
+		if err := rm.syncEntries(ctx, desired, latest); err != nil {
+			return nil, err
+		}
+		updated, err = rm.sdkFind(ctx, desired)
+		if err != nil {
+			return nil, err
+		}
```
Why is this need with Entries but not associations? maybe it useful to call it right before existing `customUpdateNetworkAcl` ?

nishant221 replied:
Done.

A-Hilaly commented on pkg/resource/network_acl/hooks.go:
```diff
@@ -0,0 +1,677 @@
+					},
+				},
+			}
+			dna_output, err := rm.sdkapi.DescribeNetworkAclsWithContext(ctx, dna_input)
```
nit: can we continue using camel case for go code? `dna_output` => `dnaOutput`. There are multiple similar cases in this file

nishant221 replied:
Fixed these.

A-Hilaly commented on pkg/resource/network_acl/hooks.go:
```diff
@@ -0,0 +1,677 @@
+			if err != nil {
+				return err
+			}
+			input.NetworkAclId = default_nacl.NetworkAcls[0].NetworkAclId
+			if len(dna_output.NetworkAcls) != 1 {
+				return errors.New("unexpected output from describenetworkacls for the given subnet")
```
nit: L300 better be called after L301

nishant221 replied:
Modified.

A-Hilaly commented on templates/hooks/network_acl/sdk_create_post_set_output.go.tpl:
```diff
@@ -0,0 +1,25 @@
+	if rm.requiredFieldsMissingForCreateNetworkAcl(&resource{ko}) {
+		return nil, ackerr.NotFound
+	}
```
Is it possible that we reach this case?

nishant221 replied:
It's unlikely but added this to cover any edge case where `CreateNetworkAcl` failed to create NACL and errors are not caught by [this](https://github.com/aws-controllers-k8s/ec2-controller/pull/140/files#diff-1fb337f82754cf5c7ea58918ef9328698030a802e4adf859fe0f7b5a01216e68R246).

A-Hilaly commented on templates/hooks/network_acl/sdk_create_post_set_output.go.tpl:
```diff
@@ -0,0 +1,25 @@
+		}
+	}
+
+	toAdd, toDelete := computeTagsDelta(desired.ko.Spec.Tags, ko.Spec.Tags)
+	if len(toAdd) == 0 && len(toDelete) == 0 {
+		// if desired tags and response tags are equal,
+		// then assign desired tags to maintain tag order
+		ko.Spec.Tags = desired.ko.Spec.Tags
+	}
```
Is this operation necessary? i see that `sdkCreate` accepts tags. Maybe it doesn't "really" create them (i have seen a similar case before :) )?

nishant221 replied:
I tested without this . I don't think we need this. Removed.

A-Hilaly reviewed (commented):
This is very good stuff!! Thank you @nishant221 !!
I have few nit/comments below

nishant221 commented:
/test ec2-kind-e2e

nishant221 commented:
@RedbackThomson / @A-Hilaly Made all the changes as per the feedbacks and tried to provide answer for the couple of questions. Also, modified e2e test to cover remaining code path as well.

Kindly have a look.

nishant221 commented:
/test ec2-kind-e2e

nishant221 commented:
/test ec2-kind-e2e

A-Hilaly reviewed (approved):
Neat!! :+1:
/approve
/cc @RedbackThomson

A-Hilaly commented:
/test all

nishant221 commented:
/test ec2-kind-e2e
